src/core/music_score_events.h

```diff
@@ -0,0 +1,668 @@
+#pragma once
+
+#include <vector>
+#include <algorithm>
+#include <unordered_map>
+#include <memory>
+
+template <typename EventType> class MusicScoreEventBrowser;
+
+
+typedef int32_t BarNumber;
+typedef int32_t BeatNumber;
+
+/**
+ * Marks a specific beat of a specific bar of music.
+ */
+struct BeatAndBar {
+	/**
+	 * Constructor.
+	 * @param bar The bar of the time that this object marks. Bar 0 is the first bar in the show.
+	 * @param beat The beat of the time that this object marks (relative to the start of the bar that it marks). Beat 0 is the first beat in the bar.
+	 */
+	BeatAndBar(BarNumber bar, BeatNumber beat) : bar(bar), beat(beat) {};
+
+	BarNumber bar;
+	BeatNumber beat;
+
+	/**
+	 * Returns whether or not this object marks the same time as the other.
+	 * @param other The time to compare to this one.
+	 * @return True if this object marks the same time as the input; false otherwise.
+	 */
+	inline bool operator==(const BeatAndBar other) const {
+		return (bar == other.bar
+			&& beat == other.beat);
+	};
+
+	/**
+	 * Returns whether or not this object marks an earlier time than the other.
+	 * @param other The time to compare to this one.
+	 * @return True if this object marks an earlier time than the input; false otherwise.
+	 */
+	inline bool operator<(const BeatAndBar other) const {
+		if (bar < other.bar) {
+			return true;
+		}
+		if (bar == other.bar) {
+			return (beat < other.beat);
+		}
+		return false;
+	};
+
+	/**
+	 * Returns whether or not this object marks a later time than the other.
+	 * @param other The time to compare to this one.
+	 * @return True if this object marks a later time than the input; false otherwise.
+	 */
+	inline bool operator>(const BeatAndBar other) const {
+		if (bar > other.bar) {
+			return true;
+		}
+		if (bar == other.bar) {
+			return (beat > other.beat);
+		}
+		return false;
+	};
```

Comment:
could be this be written as:

bool operator>(const BeatAndBar& other) const { return !(*this <= other); }